src/typeChecker/typeErrorChecker.ts

```diff
@@ -908,55 +931,80 @@ function hasTypeMismatchErrors(actualType: Type, expectedType: Type): boolean {
       if (actualType.kind === 'variable') {
-        // If both are variable types, compare without expanding;
-        // name and type arguments must match
-        if (expectedType.name !== actualType.name) {
-          return true
-        }
-        if (expectedType.typeArgs === undefined) {
-          return actualType.typeArgs?.length !== 0
-        }
-        if (actualType.typeArgs?.length !== expectedType.typeArgs.length) {
-          return true
-        }
-        for (let i = 0; i < expectedType.typeArgs.length; i++) {
-          if (hasTypeMismatchErrors(actualType.typeArgs[i], expectedType.typeArgs[i])) {
-            return true
+        // If both are variable types, types match if both name and type arguments match
+        if (expectedType.name === actualType.name) {
+          if (expectedType.typeArgs === undefined || expectedType.typeArgs.length === 0) {
+            return actualType.typeArgs === undefined ? false : actualType.typeArgs.length !== 0
           }
-        }
-        return false
-      }
-      // Expand variable type and compare expanded type with actual type
-      const aliasType = lookupTypeAlias(expectedType.name, env)
-      if (aliasType && aliasType.kind === 'forall') {
-        // Clone type to prevent modifying generic type saved in type env
-        let polyType = cloneDeep(aliasType.polyType)
-        if (expectedType.typeArgs) {
-          if (aliasType.typeParams?.length !== expectedType.typeArgs.length) {
+          if (actualType.typeArgs?.length !== expectedType.typeArgs.length) {
             return true
           }
           for (let i = 0; i < expectedType.typeArgs.length; i++) {
-            polyType = substituteVariableTypes(
-              polyType,
-              aliasType.typeParams[i],
-              expectedType.typeArgs[i]
-            )
+            if (
+              hasTypeMismatchErrors(
+                node,
+                actualType.typeArgs[i],
+                expectedType.typeArgs[i],
+                visitedTypeAliasesForActualType,
+                visitedTypeAliasesForExpectedType,
+                skipTypeAliasExpansion
+              )
+            ) {
+              return true
+            }
           }
+          return false
         }
-        return hasTypeMismatchErrors(actualType, polyType)
       }
-      return true
+      for (const visitedType of visitedTypeAliasesForExpectedType) {
+        if (visitedType.name === expectedType.name) {
+          // Circular dependency, treat as type mismatch
+          return true
+        }
+      }
+      // Skips expansion, treat as type mismatch
+      if (skipTypeAliasExpansion) {
+        return true
+      }
+      // Pair and List should not be added to visited type aliases
+      // as they can be used repeatedly without causing circular dependencies
+      if (expectedType.name !== 'Pair' && expectedType.name !== 'List') {
+        visitedTypeAliasesForExpectedType.push(expectedType)
+      }
```

Comment:
Does this mean that I can't declare my own `Triple` type and use it like `Triple<number, string, Triple<number, string, boolean>>`?

Reply:
Thanks for pointing this out! Will need to tighten the circular dependency check to take into account the type arguments as well, after which your stated example should be possible.